datafusion/physical-plan/src/aggregates/group_values/column.rs

```diff
@@ -75,55 +176,317 @@ pub struct GroupValuesColumn {
+        }
+    }
+
+    fn scalarized_intern(
+        &mut self,
+        cols: &[ArrayRef],
+        batch_hashes: &[u64],
+        groups: &mut Vec<usize>,
+    ) {
+        if self.scalarized_indices.is_empty() {
+            return;
+        }
+
+        let mut map = mem::take(&mut self.map);
+
+        for &row in &self.scalarized_indices {
+            let target_hash = batch_hashes[row];
+            let entry = map.get_mut(target_hash, |(exist_hash, group_index_view)| {
+                // Somewhat surprisingly, this closure can be called even if the
+                // hash doesn't match, so check the hash first with an integer
+                // comparison first avoid the more expensive comparison with
+                // group value. https://github.com/apache/datafusion/pull/11718
+                target_hash == *exist_hash
+            });
+
+            // Only `rows` having the same hash value with `exist rows` but different value
+            // will be process in `scalarized_intern`.
+            // So related `buckets` in `map` is ensured to be `Some`.
+            let Some((_, group_index_view)) = entry else {
+                unreachable!()
+            };
+
+            // Perform scalarized equal to
+            if self.scalarized_equal_to(&group_index_view, cols, row, groups) {
+                // Found the row actually exists in group values,
+                // don't need to create new group for it.
+                continue;
+            }
+
+            // Insert the `row` to `group_values` before checking `next row`
+            let group_idx = self.group_values[0].len();
+            let mut checklen = 0;
+            for (i, group_value) in self.group_values.iter_mut().enumerate() {
+                group_value.append_val(&cols[i], row);
+                let len = group_value.len();
+                if i == 0 {
+                    checklen = len;
+                } else {
+                    debug_assert_eq!(checklen, len);
+                }
+            }
+
+            // Check if the `view` is `inlined` or `non-inlined`
+            if group_index_view.is_non_inlined() {
+                // Non-inlined case, get `group_index_list` from `group_index_lists`,
+                // then add the new `group` with the same hash values into it.
+                let list_offset = group_index_view.value() as usize;
+                let group_index_list = &mut self.group_index_lists[list_offset];
+                group_index_list.push(group_idx);
+            } else {
+                // Inlined case
+                let list_offset = self.group_index_lists.len();
+
+                // Create new `group_index_list` including
+                // `exist group index` + `new group index`.
+                // Add new `group_index_list` into ``group_index_lists`.
+                let exist_group_index = group_index_view.value() as usize;
+                let new_group_index_list = vec![exist_group_index, group_idx];
```

Comment:
Do we need `exist_group_index`? I think it is the one that has the same value and different value for other rows, so we don't need to check the value once again.

Reply:
The `map` is a bit different with origin.

The original `map` can be seen as `HashMap<(hash value, group value), ()>`, and each bucket only points to `single group`.

And cureent `map` is actullay like `HashMap<hash value,  Vec<group value>>`, each bucket will point to a list of `groups with the same hash value(hash collision)`.

In this branch, the `new group with the same hash value` found, so we need to switch `inlined group view(the actual and single group index)` to `non-inlined group view(a index to gourp indices list)`.

Reply:
I have added comments to explain the new `hashtable` in `VectorizedGroupValuesColumn`.